### ISSUE TITLE
Build issue: Where to get msgpack headers

### ISSUE THREAD

**Issue by petten** [darrenjs/wampcc#22]
Need to add:

`include_directories("${PROJECT_SOURCE_DIR}/3rdparty/msgpack-c/include")`

around (above or below):

`include_directories("${PROJECT_SOURCE_DIR}/3rdparty/websocketpp")`

in the main `CMakeLists.txt` file to avoid picking up external `msgpack` headers from (under):

`include_directories(${INSTALL_INC_DIR})`

which comes after the internal `3rdparty` lines above.

**Comment by darrenjs** [comment]
So the issue is that, for your cmake based build, you were finding headers outside of the wampcc directories?  I.e., msgpack headers unrelated to the wampcc project ?

**Comment by petten** [comment]
Yes, and yes.

I've got a different version of `msgpack-c`, which is a sibling of `wampcc`, directory structure-wise. Both are built and installed to the same place. That separate `msgpack-c` is used by some other libraries and user code, and is not the same as the `msgpack-c`  under the `3rdparty` subdirectory of `wampcc`. (Got a compilation error.)

I thought your having your own version of `msgpack-c` under the `3rdparty` subdirectory was for this exact reason - to avoid referring to an independent but incompatible version of `msgpack-c`.

See, you have one `include_directories()` line for `websocketpp`, which makes the `wampcc` code refer to the `websocketpp` headers under the `3rdparty` subdirectory. You have to do the same for `msgpack-c`.

**Comment by petten** [comment]
BTW, I also have a different version of `websocketpp` as a sibling of both that `msgpack-c` and `wampcc`.

The existing `include_directories(... websocketpp)` line does the job of making the `wampcc` code refer to your `3rdparty` version.

**Comment by darrenjs** [comment]
Hi, I'm just looking at this issue.

When I compile using cmake (lets say I have made a change to libs/json/msgpack_serialiser.cc), here is the cmake compile line:

```shell
/usr/bin/c++ -I/home/buildusr/src/wampcc_build_cmake -I/home/buildusr/src/wampcc/include -I/home/buildusr/src/wampcc/3rdparty/websocketpp -I/home/buildusr/src/wampcc/3rdparty -I/home/buildusr/src/wampcc/3rdparty/msgpack-c/include -I/home/buildusr/opt/jansson-2.10/include  -Wall   -std=gnu++11 -o CMakeFiles/wampcc_json_static.dir/msgpack_serialiser.cc.o -c /home/buildusr/src/wampcc/libs/json/msgpack_serialiser.cc
```


... so expanding this:

```
01: /usr/bin/c++
02: -I/home/buildusr/src/wampcc_build_cmake
03: -I/home/buildusr/src/wampcc/include
04: -I/home/buildusr/src/wampcc/3rdparty/websocketpp
05: -I/home/buildusr/src/wampcc/3rdparty
06: -I/home/buildusr/src/wampcc/3rdparty/msgpack-c/include
07: -I/home/buildusr/opt/jansson-2.10/include
08: -Wall   -std=gnu++11 -o CMakeFiles/wampcc_json_static.dir/msgpack_serialiser.cc.o -c /home/buildusr/src/wampcc/libs/json/msgpack_serialiser.cc
```

So at line 06, there is already this dedicated include directive for the msgpack-c headers.  This is produced by the libs/json/CMakeList.txt

So in your error case, what file was being compiled at the time you encountered the error?  Could you post an equivalent sample output?

**Comment by petten** [comment]
OK, sorry. I changed the cmake file a bit to adapt to our environment, where headers for all `3rdparty` libraries (including `wampcc`) are installed under `INSTALL_INC_DIR`. Libraries such as `uv` are built ahead of `wampcc`, so that `wampcc` can refer to the dependencies. (Do not confuse the above-mentioned `3rdparty` with the one under `wampcc`.) So I have:

`include_directories("${PROJECT_BINARY_DIR}")`
`include_directories("${PROJECT_SOURCE_DIR}/include")`
`include_directories("${PROJECT_SOURCE_DIR}/3rdparty/msgpack-c/include")`
`include_directories("${PROJECT_SOURCE_DIR}/3rdparty/websocketpp")`
`include_directories("${PROJECT_SOURCE_DIR}/3rdparty")`
`include_directories(${INSTALL_INC_DIR})`
`include_directories(${INSTALL_INC_DIR}/uv)`

and I had the "`Try to find libuv`" and "`Try to find jansson`", etc., sections (before the above) completely commented out, because they are located under `INSTALL_INC_DIR`, where one can also find `msgpack` and `websocketpp` headers from our separately imported copies in parallel to `wampcc`. (I.e., we have:

`+ 3rdparty`
`+ 3rdparty/....`
`+ 3rdparty/jansson`
`+ 3rdparty/uv`
`+ 3rdparty/msgpack`
`+ 3rdparty/websocketpp`
`+ 3rdparty/wampcc`
`+ 3rdparty/....`

when doing a directory listing.)

I am not sure how you got the line

`06: -I/home/buildusr/src/wampcc/3rdparty/msgpack-c/include`

because the main cmake file in both the current `wampcc` master and release 1.5 has only:

`include_directories("${PROJECT_BINARY_DIR}")`
`include_directories("${PROJECT_SOURCE_DIR}/include")`
`include_directories("${PROJECT_SOURCE_DIR}/3rdparty/websocketpp")`
`include_directories("${PROJECT_SOURCE_DIR}/3rdparty")`

You have:

`07: -I/home/buildusr/opt/jansson-2.10/include`

because it was found earlier (before the `include_directories` lines).

And lines `02-05` correspond to the four `include_directories` above.

In our case, when building `wampcc`, it's possible to pick up headers from the incompatible `msgpack` (a sibling of `wampcc` as in the directory structure above) installed under the `INSTALL_INC_DIR`, if the line

`include_directories("${PROJECT_SOURCE_DIR}/3rdparty/msgpack-c/include")`

is not present.

**Comment by darrenjs** [comment]
Hi,

The line `06` appears when I compile the `libs/json/msgpack_serialiser.cc` file (for example, make a change to that file, then run the compilation again).

The line is defined in libs/json/CMakeLists.txt : 

```
  target_include_directories(wampcc_json_static
                             PRIVATE ${PROJECT_SOURCE_DIR}/3rdparty/msgpack-c/include
                             ${JANSSON_INCLUDE_DIRS} )
```

... this makes sense, because that include is only needed during the compilation of the wampcc_json library.   So it doesn't need to appear in the main CMakeLists.txt file.

**Comment by darrenjs** [comment]
Also, fyi, this is how I typically manage my thirdparty libraries.  I do this for the wampcc project, but also for all other C++ projects I work on.

I define a root directory where all 3rd party libraries that I built will be installed.  Lets call this `/opt/libs`, but can be anywhere.

Under there, I have a separate directory for each project/version that I build & install.  E.g.

```
/opt/libs/jansson-2.10
/opt/libs/libuv-1.10.2
/opt/libs/wampcc-1.5
/opt/libs/cppcheck-1.68
/opt/libs/libev-4.22
```
Under each of these directories are the typical `include` and `lib` associated with projects that have an install stage.  Sometimes under each of these I might also have a `src` and `build` directory, which is where I would unpack the source code and perform the build.  However the source-unpack and src build can be performed anywhere.

**Comment by darrenjs** [comment]
Also, fyi, this is how I typically manage my thirdparty libraries. I do this for the wampcc project, but also for all other C++ projects I work on.

I define a root directory where all 3rd party libraries that I built will be installed. Lets call this `/opt/libs`, but can be anywhere.

Under there, I have a separate directory for each project/version that I build & install. E.g.

```
/opt/libs/jansson-2.10
/opt/libs/libuv-1.10.2
/opt/libs/wampcc-1.5
/opt/libs/cppcheck-1.68
/opt/libs/libev-4.22
```

Under each of these directories are the typical `include` and `lib` directories associated with projects that have an install stage. Sometimes under each of these I might also have a `src` and `build` directory, which is where I would unpack the source code and perform the build. However the source-unpack and  build can be performed anywhere.

**Comment by petten** [comment]
The reason it does not work for me is that my custom-added `$INSTALL_INC_DIR` in the main `cmake` file is ahead of the `target_include_directories(...msgpack...)` appended from the `libs/json` subdirectory.

Need to find a way to prepend the `target` ones to those added in the main `cmake` file.

**Comment by petten** [comment]
I think this is asking too much from you. So I'm closing it.

I basically can either do what I already did (i.e., adding `include msgpack` in the main `cmake` file ahead of `${INSTALL_INC_DIR}`) or add `${INSTALL_INC_DIR}` and/or `${INSTALL_INC_DIR}/uv` to the `target_include...` line in the `cmake` file of respective subdirectories that require either or both of them.

**Comment by darrenjs** [comment]
Hi ... reviewing this again.  I notice you say  that you changed the makefile to reflect your installation layout, and then you have this
```
include_directories(${INSTALL_INC_DIR})
include_directories(${INSTALL_INC_DIR}/uv)
```

... I think that is the problem.  During normal wampcc compilation, it should not need to look there.  Why did you need to add that?  If you want to include the path to libuv, you can provide that as an option when calling cmake for this first time.

```
cmake --verbose  -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON \
-DLIBUV_DIR=/opt/libuv-1.10.2 \
-DJANSSON_DIR=/opt/jansson-2.10 \
-DCMAKE_INSTALL_PREFIX=/var/tmp/wampcc_install \
/var/tmp/wampcc_src
```

**Comment by petten** [comment]
In my case, all libraries (even including `openssl`; except only for the most fundamental ones such as `pthread` and whatever languages' standard libraries) are treated the same as `wampcc`.

They are all built as siblings of `wampcc` and their respective headers installed under `${INSTALL_INC_DIR}`. Those libraries that are dependencies of others are built (and therefore installed) first (e.g., `libuv`), hence can be referred to by the ones that are built later (e.g., `wampcc`, referring to `libuv` built earlier).

I think the primary goal of this arrangement is that when everything is distributed a greater degree of compatibility is ensured. Because you avoid the libraries installed by the user himself in the standard locations that may be incompatible with your applications. (This, of course, requires that the user insert the path to your libraries ahead of his usual ones when running your software.)

In short, we don't add locations of libraries like `libuv` and `jansson` on the `cmake` command line, because they are all supposed to be under `${INSTALL_INC_DIR}` (for headers, and those other artifacts in their respective install directories).

**Comment by petten** [comment]
Darren, may I ask a separate question here? (I'm not sure whether there is some kind of a discussion forum for individual projects here @github or somewhere else.)

Previously I asked you about whether as a callee a `wampcc_session` supported multiple threads. The answer was no.

Now I noticed a `request_id` in the `invocation`  argument to the callback, and it must be passed as an argument to the eventual `yield()` call sending the response. Does it mean that as long as I keep track of both the `wampcc_session` reference and the `request_id` for each incoming call, the eventual `yield()` calls sending the respective responses can be out of order?

Individual incoming calls may require different amounts of processing time. The thinking is that the incoming calls can be dispatched to a thread pool through an inbound queue. Those threads do their work and when ready push their respective results into an outbound queue. When the `wampcc` callee's callback logic flow reaches the point of calling `yield()`, it dequeues from the outbound queue (wait until at least one is ready) to send a response, which may or may not correspond to the incoming request during the same call.

Furthermore, can we call `yield()` more than once (as long as there are more items in the outbound queue) for a single incoming call?

**Comment by darrenjs** [comment]
Hi.  I'm replying to your discussion on your deployment pattern.

Building all of your libraries (openssl, libuv, jansson, wampcc etc) under a single location is a good practice I think, because it does make it easier to deploy your application to other machines. That's what I also do; I typically put everything I manually build under `/opt` , and then later I can copy that entire tree to another box.  I'm on the same page as you there.

However, open source libraries based on autotools & cmake can actually support that arrangement via their command line options that allow the builder to say where to find the dependent libraries.  For example, with wampcc, at  automake/cmake configuration stage, you just tell it "all my libraries are under /opt", and it finds them there etc.  That's better than modifying the make file to alter paths directly.

For your threading question, maybe open new issue?
